### DISCUSSION TITLE
determine if path between two nodes that contain another node?

### DISCUSSION THREAD

**Discussion by adam2392** [networkx/networkx#5829] [Q&A]
Is it possible to determine if an adjacency path exists between two nodes that contains another node?

E.g.

`path_with_node(start_node, end_node, check_node) -> bool`

would return True if there is an acyclic path from `start_node` to `end_node` with `check_node` in the path and False if no such paths exist?

**Comment by MridulS** [comment]
There is no inbuilt function for this (I think!) but this could be implemented for a general Graph with something like:
``` python
# NOT TESTED!
def path_with_node(G, start_node, end_node, check_node):
    try:
        sp = nx.shortest_path(G, start_node, end_node)
    except nx.NetworkXNoPath:
        return False
    return check_node in sp
```
PS: Does the path need to be acyclic (i.e. unique) or the graph needs to be acyclic?

**Reply by adam2392** [reply]
The graph is assumed a cyclic. 

I think that function wouldn't work tho since if there's a longer path with Check_node this would not work. 

I suppose it could also work by creating a generator over all possible paths between start and end and then determining if check_node is part of any of those paths. I think this occurs via DFS? However im wondering if there's something more efficient in networkx?

**Reply by dschult** [reply]
I think the way to approach this is to find a path from `start_node` to `check_node`. Then find a path from `check_node` to `end_node`.   So code like this should work:
```python
if G.has_path(start_node, check_node) and G.has_path(check_node, end_node):
```

**Reply by adam2392** [reply]
But what if checknode to end node goes through start node? Then it would be invalid :/

**Reply by dschult** [reply]
Does that mean you can't have any nodes overlap between the two paths?
Unfortunately there are combinatorially many versions of this problem. You will need to be very careful how to state it. :}

**Reply by dschult** [reply]
Another perhaps helpful function is `nx.all_simple_paths`.

**Reply by adam2392** [reply]
Ah okay. So what I would like is to say among all non cycle paths (that don't say go A-B-A) is there a path from "start" to "end" that goes through "check"?

If so return True. I get that worst case complexity would need to go through all paths from Start to end. Is there an average case improvement tho over:

- iterate over paths from Start to end without repeating elements (using what's in find_simple_paths)
- if at any point check is in the path, return True
- return False

? I.e. is it plausible that BFS does better? Since In BFS case you could hone in on all paths that touch "check" right away and then maybe do DFS to see if they lead to dead end. 

My big O on graphs is bad so maybe this is overthinking it and I need to just iterate over possibly all paths?

**Reply by dschult** [reply]
My intuition may not be very good for this problem, but it says that BFS and/or shortest_path are definitely the way to go here. You could even try a bidirectional BFS alternating forward steps from `start` with backward steps from end, looking for `check` in the intersection. It could stop one direction once check is found--I think that would prevent cycles... There is a `nx.bidirectional_shortest_path` you could take a hack at. :}

**Reply by boothby** [reply]
Suppose you 
1. find the shortest path from the start node to the check node in the graph obtained by deleting the end node,
2. look for the shortest path from the check node to the end node in the graph obtained by deleting the path prefix (all the nodes but the check node) found in step (1)

If step (2) fails, can any permissible path exist?  I haven't thought on this much and I'm just heading to bed, but my intuition is that this works fine for undirected graphs unless the check node is on either side of a 1-node separator between the start and end nodes (and there's no path to be found)... and that this probably won't help in an directed graph.

**Reply by boothby** [reply]
The above is quite wrong but looking at cuts is a decent approach.  At least, for undirected graphs, one can look for node-disjoint paths from the check node to a special target node that the start and end node are connected to...

https://networkx.org/documentation/stable/reference/algorithms/generated/networkx.algorithms.connectivity.disjoint_paths.node_disjoint_paths.html
